File: go/core/ast.go
package core

// AST node types produced by the scanner/structurer — port of ast.ts / ast.rs.
// Pure data; the sealed Block/TableOrFence unions become a Go interface with a
// closed set of implementers (type-switched in the conformance projection).
// Immutability is by convention — fields are never mutated after construction.

// SegmentOffset maps a block-text offset to its source offset (both UTF-16).
type SegmentOffset struct {
	TextOffset   int
	SourceOffset int
}

// Heading is a markdown heading (`#`..`######`); Level is 1–6.
type Heading struct {
	Level int
	Text  string
	Span  Span
}

// Paragraph is a markdown paragraph.
type Paragraph struct {
	Text       string
	Span       Span
	SegmentMap []SegmentOffset
}

// ListItem is a single list item (`-`/`*` or numbered).
type ListItem struct {
	Text       string
	Span       Span
	SegmentMap []SegmentOffset
	Ordered    bool
	MarkerSpan Span
}

// Blockquote is a markdown blockquote (`>`).
type Blockquote struct {
	Text       string
	Span       Span
	SegmentMap []SegmentOffset
}

// Row is one row of a table: Cells and CellSpans are parallel, same-length.
type Row struct {
	Cells     []string
	CellSpans []Span
	Span      Span
}

// Table is a markdown table: a header Row plus zero or more data rows.
type Table struct {
	Span   Span
	Header Row
	Rows   []Row
}

// Fence is a fenced code block; Info is the text after the opening fence.
type Fence struct {
	Span     Span
	Info     string
	Body     string
	BodySpan Span
}

// ThematicBreak is a thematic break (`---`/`***`/`___`).
type ThematicBreak struct {
	Span Span
}

// Block is a markdown block node — the closed union the structurer matches over.
// Only the seven types in this file implement it.
type Block interface {
	blockSpan() Span
}

func (h Heading) blockSpan() Span       { return h.Span }
func (p Paragraph) blockSpan() Span     { return p.Span }
func (l ListItem) blockSpan() Span      { return l.Span }
func (b Blockquote) blockSpan() Span    { return b.Span }
func (t Table) blockSpan() Span         { return t.Span }
func (f Fence) blockSpan() Span         { return f.Span }
func (t ThematicBreak) blockSpan() Span { return t.Span }

// blockSpanOf returns a block's source span.
func blockSpanOf(b Block) Span { return b.blockSpan() }

// Example is one matched example: the heading scope above it (outer→inner) plus
// its body blocks (first is the candidate primary block, then any attachments).
type Example struct {
	ScopeStack []string
	Span       Span
	Body       []Block
}

// VarDoc is a parsed source file: its matched examples plus unattached
// table/fence blocks (each of which is a Table or Fence).
type VarDoc struct {
	Path              string
	Source            string
	Examples          []Example
	OrphanAttachments []Block
}

File: rust/core/src/ast.rs
//! AST node types produced by the scanner/structurer — port of `ast.ts` /
//! `Ast.java`. Pure data; the sealed `Block`/`TableOrFence` interfaces become
//! Rust enums (exhaustive `match` replaces `instanceof`). Immutability is by
//! construction (owned fields, no mutation) — Java's `List.copyOf` defensive
//! copies have no Rust analog.

use crate::span::Span;

/// Maps a block-text offset to its source offset (both UTF-16).
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub struct SegmentOffset {
    pub text_offset: usize,
    pub source_offset: usize,
}

impl SegmentOffset {
    pub fn new(text_offset: usize, source_offset: usize) -> SegmentOffset {
        SegmentOffset {
            text_offset,
            source_offset,
        }
    }
}

/// A markdown heading (`#`..`######`); `level` is 1–6.
#[derive(Clone, Debug, PartialEq, Eq)]
pub struct Heading {
    pub level: usize,
    pub text: String,
    pub span: Span,
}

/// A markdown paragraph.
#[derive(Clone, Debug, PartialEq, Eq)]
pub struct Paragraph {
    pub text: String,
    pub span: Span,
    pub segment_map: Vec<SegmentOffset>,
}

/// A single list item (`-`/`*` or numbered).
#[derive(Clone, Debug, PartialEq, Eq)]
pub struct ListItem {
    pub text: String,
    pub span: Span,
    pub segment_map: Vec<SegmentOffset>,
    pub ordered: bool,
    pub marker_span: Span,
}

/// A markdown blockquote (`>`).
#[derive(Clone, Debug, PartialEq, Eq)]
pub struct Blockquote {
    pub text: String,
    pub span: Span,
    pub segment_map: Vec<SegmentOffset>,
}

/// One row of a table: `cells` and `cell_spans` are parallel, same-length.
#[derive(Clone, Debug, PartialEq, Eq)]
pub struct Row {
    pub cells: Vec<String>,
    pub cell_spans: Vec<Span>,
    pub span: Span,
}

/// A markdown table: a header [`Row`] plus zero or more data rows.
#[derive(Clone, Debug, PartialEq, Eq)]
pub struct Table {
    pub span: Span,
    pub header: Row,
    pub rows: Vec<Row>,
}

/// A fenced code block; `info` is the text after the opening fence.
#[derive(Clone, Debug, PartialEq, Eq)]
pub struct Fence {
    pub span: Span,
    pub info: String,
    pub body: String,
    pub body_span: Span,
}

/// A thematic break (`---`/`***`/`___`).
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub struct ThematicBreak {
    pub span: Span,
}

/// A markdown block node — the closed union the structurer matches over.
#[derive(Clone, Debug, PartialEq, Eq)]
pub enum Block {
    Heading(Heading),
    Paragraph(Paragraph),
    ListItem(ListItem),
    Blockquote(Blockquote),
    Table(Table),
    Fence(Fence),
    ThematicBreak(ThematicBreak),
}

impl Block {
    /// The block's source span (exhaustive over the union).
    pub fn span(&self) -> Span {
        match self {
            Block::Heading(h) => h.span,
            Block::Paragraph(p) => p.span,
            Block::ListItem(l) => l.span,
            Block::Blockquote(b) => b.span,
            Block::Table(t) => t.span,
            Block::Fence(f) => f.span,
            Block::ThematicBreak(t) => t.span,
        }
    }
}

/// The block kinds that may appear as a [`VarDoc`] orphan attachment (`Table | Fence`).
#[derive(Clone, Debug, PartialEq, Eq)]
pub enum TableOrFence {
    Table(Table),
    Fence(Fence),
}

/// One matched example: the heading scope above it (outer→inner) plus its body
/// blocks (first is the candidate primary block, then any trailing attachments).
#[derive(Clone, Debug, PartialEq, Eq)]
pub struct Example {
    pub scope_stack: Vec<String>,
    pub span: Span,
    pub body: Vec<Block>,
    /// True when a heading or thematic break (`---`) sits between this candidate
    /// and the previous one — i.e. a syntactic delimiter separates them (also
    /// true for the first candidate). The planner uses it to decide grouping: a
    /// matching candidate with this false merges into the open example rather
    /// than starting a new one. See ADR 0012.
    pub preceded_by_delimiter: bool,
}

/// A parsed source file: its matched examples plus unattached table/fence blocks.
#[derive(Clone, Debug, PartialEq, Eq)]
pub struct VarDoc {
    pub path: String,
    pub source: String,
    pub examples: Vec<Example>,
    pub orphan_attachments: Vec<TableOrFence>,
}
